File: android/src/main/kotlin/com/usercentrics/sdk/flutter/api/UsercentricsProxy.kt
package com.usercentrics.sdk.flutter.api

import android.content.Context
import com.usercentrics.sdk.Usercentrics
import com.usercentrics.sdk.UsercentricsOptions
import com.usercentrics.sdk.UsercentricsReadyStatus
import com.usercentrics.sdk.UsercentricsSDK
import com.usercentrics.sdk.errors.UsercentricsError

internal interface UsercentricsProxy {
    val instance: UsercentricsSDK

    fun initialize(context: Context?, options: UsercentricsOptions)

    fun isReady(
        onSuccess: (UsercentricsReadyStatus) -> Unit,
        onFailure: (UsercentricsError) -> Unit
    )
}

internal object UsercentricsProxySingleton : UsercentricsProxy {

    override val instance: UsercentricsSDK
        get() = Usercentrics.instance

    override fun initialize(context: Context?, options: UsercentricsOptions) {
        context ?: return

        Usercentrics.initialize(context, options)
    }

    override fun isReady(
        onSuccess: (UsercentricsReadyStatus) -> Unit,
        onFailure: (UsercentricsError) -> Unit
    ) = Usercentrics.isReady(onSuccess, onFailure)
}

File: ios/usercentrics_sdk/Sources/usercentrics_sdk/API/UsercentricsProxy.swift
import Foundation
import Usercentrics
import Flutter

protocol UsercentricsProxyProtocol {
    var shared: UsercentricsSDK { get }
    func isReady(onSuccess: @escaping ((UsercentricsReadyStatus) -> Void), onFailure: @escaping ((Error) -> Void))
    func configure(options: UsercentricsOptions)
}

struct UsercentricsProxy: UsercentricsProxyProtocol {
    var shared: UsercentricsSDK {
        UsercentricsCore.shared
    }

    func isReady(onSuccess: @escaping ((UsercentricsReadyStatus) -> Void), onFailure: @escaping ((Error) -> Void)) {
        UsercentricsCore.isReady(onSuccess: onSuccess, onFailure: onFailure)
    }

    func configure(options: UsercentricsOptions) {
        UsercentricsCore.configure(options: options)
    }
}
